### DISCUSSION TITLE
Using frames

### DISCUSSION THREAD

**Discussion by avidseeker** [libjxl/libjxl#3900] [Q&A]
According to [JPEG XL spec](https://github.com/ImageMagick/jpeg-xl/blob/main/doc/format_overview.md#frames)

> Zero-duration frames are possible and represent different layers of the image.

Is this featured supported on GIMP or Inkscape or even a commandline tool like imagemagick?

E.g usage: how can I list all layers within a jxl file? How do I merge separate images 1.jxl, 2.jxl, ... into layer1 layer2, ... of out.jxl

Stack exchange question: https://graphicdesign.stackexchange.com/questions/166219/

**Comment by alistair7** [comment]
The Gimp plugin (and probably most other tools) always "coalesces" the JXL into full-sized pre-blended animation frames, so it won't show you the zero-duration layers within those frames.  This is mainly because JXL frames can reference each other in complicated ways that aren't supported by most editors.  e.g. there's no way to specify in Gimp that frame 5 should be added to frame 1, so the plugin plays it safe and just lets the decoder do the required blending and shows the result as a simple "replace" frame.

Hacking the Gimp plugin to disable coalescing is an option, with the caveat that "fancy" blending might not work correctly:
https://github.com/libjxl/libjxl/blob/adeeccc415f7c18e6a1732d4c9d6d905ed8e2f9b/plugins/gimp/file-jxl-load.cc#L104-L109

Listing the frames/layers can be done with the `jxlinfo` tool that comes with libjxl.

For merging frames together you could try
https://github.com/kampidh/JXL-Frame-Stitcher
(I haven't tried it, but by all accounts it's very effective.)

Outside of that, you would probably have to write your own code to use the libjxl API directly.

I've made my own half-baked command-line tool that does something like what you're attempting.  It can split a JXL file into individual JXL frames with an accompanying JSON file specifying the frame offsets, blend modes, durations, etc., and it can read this back in to re-merge it again.  I was considering putting it on GitHub... but I'd need to find the time to make it fit for human consumption, as it was only ever meant for the specific thing I was doing at the time, so it's not very generic :).  It also requires re-encoding the frames with every merge/split operation, so it's only really useful for lossless files.
